### PR TITLE
Fix bug in SmoothRateLimiter where executionDelay is not honored

### PR DESCRIPTION
**Pull request by lesterhaynes**
Previous code tends to not actually introduce randomness before executing callbacks. This is caused by event loop iterations being rapidly scheduled back to back (such as setting the rate), which flip the boolean that tracks whether or not a delay should be incurred. This patch replaces the hasDelayed boolean with a more durable timestamp to compare against.

Most important change is a robust unit test that should accurately simulate production. Let me know if I'm missing something with this test setup. The new test fails spectacularly without the patch to SmoothRateLimiter applied.

### CODE REVIEW

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java [hunk 1]
```diff
@@ -243,10 +243,16 @@ private void updateWithNewRate()
       // before entering the next period
       _permitAvailableCount = Math.max(rate.getEvents() - (_permitsInTimeFrame - _permitAvailableCount), 0);
       _permitsInTimeFrame = rate.getEvents();
+      _delayUntil = _clock.currentTimeMillis() + _executionTracker.getNextExecutionDelay(_rate);
 
       loop();
     }
 
+    private int getRemainingTimeInPeriod()
```
Comment (rickzx) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java:251]: Is this function used anywhere?

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java [hunk 2]
```diff
@@ -317,12 +318,12 @@ public void loop()
         try
         {
           // avoids executing too many duplicate tasks
-          long nextRunRelativeTime = Math.max(0, _permitTime + rate.getPeriod() - _clock.currentTimeMillis());
-          long nextRunAbsolute = _clock.currentTimeMillis() + nextRunRelativeTime;
-          if (_nextScheduled > nextRunAbsolute || _nextScheduled <= _clock.currentTimeMillis())
+          // reschedule next iteration of the event loop to the next delay, or the beginning of the next period
+          long nextRunRelativeTime = _permitAvailableCount > 0 ? _delayUntil - now : Math.max(0, _permitTime + rate.getPeriod() - now);
+          long nextRunAbsolute = now + nextRunRelativeTime;
```
Comment (rickzx) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java:323]: This looks a bit confusing to me. How about keep nextRunRelativeTime as is and change
`long nextRunAbsolute = _permitAvailableCount > 0 ? _delayUntil : now + nextRunRelativeTime`

Comment (lesterhaynes) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java:323]: The suggested change results in EventLoop failing to honor the delay, with all unit tests failing in TestConstantQpsRateLimiter, though all tests in TestSmoothRateLimiter continue to pass. We need to augment both nextRunRelative and nextRunAbsolute with the delay in order for the next iteration of the loop to be scheduled with the delay in mind.

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java (hunk 1 above)
Comment (lesterhaynes) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java:251]: Thanks, should remove it.

**Review by rickzx — APPROVED** [review]

**Pull request merged**